spanner/sessionclient.go

```diff
@@ -230,11 +254,39 @@ func (sc *sessionClient) executeBatchCreateSessions(client *vkit.Client, createC
 			consumer.sessionCreationFailed(ToSpannerError(ctx.Err()), remainingCreateCount)
 			break
 		}
+		var mdForGFELatency metadata.MD
 		response, err := client.BatchCreateSessions(ctx, &sppb.BatchCreateSessionsRequest{
 			SessionCount:    remainingCreateCount,
 			Database:        sc.database,
 			SessionTemplate: &sppb.Session{Labels: labels},
-		})
+		}, gax.WithGRPCOptions(grpc.Header(&mdForGFELatency)))
+
+		if GFELatencyMetricsEnabled && mdForGFELatency != nil {
+			_, instance, database, errGFE := parseDatabaseName(sc.database)
+			if errGFE != nil {
+				trace.TracePrintf(ctx, nil, "Error getting instance and database name: %v", errGFE)
+				consumer.sessionCreationFailed(ToSpannerError(errGFE), remainingCreateCount)
+				break
+			}
+			// Errors should not prevent initializing the session pool.
+			ctxGFE, errGFE := tag.New(ctx,
+				tag.Upsert(tagKeyClientID, sc.id),
+				tag.Upsert(tagKeyDatabase, database),
+				tag.Upsert(tagKeyInstance, instance),
+				tag.Upsert(tagKeyLibVersion, version.Repo),
+			)
+			if errGFE != nil {
+				trace.TracePrintf(ctx, nil, "Error in adding tags in BatchCreateSessions for GFE Latency: %v", errGFE)
+				consumer.sessionCreationFailed(ToSpannerError(errGFE), remainingCreateCount)
+				break
+			}
+			errGFE = captureGFELatencyStats(ctxGFE, mdForGFELatency, "executeBatchCreateSessions")
+			if errGFE != nil {
+				trace.TracePrintf(ctx, nil, "Error in Capturing GFE Latency and Header Missing count. Try disabling and rerunning. Error: : %v", errGFE)
+				consumer.sessionCreationFailed(ToSpannerError(errGFE), remainingCreateCount)
+				break
```

Comment:
Same as above; this can also be removed if the error check for session creation is moved to before this entire block.